enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,335 @@
+
+*What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.*
+
+*How will security be reviewed and by whom? How will UX be reviewed and by whom?*
+
+## Design Details
+
+### Open Questions
+
+1. How will the user specify custom configuration, such as installation disk, static IPs?
+2. Number of revisions for the control plane - do we want to make changes to the bootstrap static pods to make them closer to the final ones?
+3. What do we do with the CRDs that are generated by operators instead of installed from manifests managed by the CVO?
+
+### Bootable installation artifact (future work)
+
+In order to embed the bootstrap-in-place-for-live-iso Ignition config to the liveCD the user need to get the liveCD and the coreos-installer binary.
+We consider adding `openshift-install create single-node-iso` command that that result a liveCD with the bootstrap-in-place-for-live-iso.ign embeded.
+It can also take things like additional manifests for setting the RT kernel (and kernel args) via MachineConfig as well
+ as supporting injecting network configuration as files and choosing the target disk drive for coreos-installer.
+Internally, create single-node-iso would compile a single-node-iso-target.yaml into Ignition (much like coreos/fcct)
+ and include it along with the Ignition it generates and embed it into the ISO.
+
+### Limitations
+
+While most CRDs get created by CVO some CRDs are created by the operators, since during the bootstrap phase there is no schedulable node, 
+ operators can't run, these CRDs won't be created until the node pivot to become the master node. 
+This imposes a limitation on the user when specifying custom manifests prior to the installation. 
+
+These are the CRDs that are not present during bootstrap:
+* clusternetworks.network.openshift.io
+* controllerconfigs.machineconfiguration.openshift.io
+* egressnetworkpolicies.network.openshift.io
+* hostsubnets.network.openshift.io
+* ippools.whereabouts.cni.cncf.io
+* netnamespaces.network.openshift.io
+* network-attachment-definitions.k8s.cni.cncf.io
+* overlappingrangeipreservations.whereabouts.cni.cncf.io
+* volumesnapshotclasses.snapshot.storage.k8s.io
+* volumesnapshotcontents.snapshot.storage.k8s.io
+* volumesnapshots.snapshot.storage.k8s.io
+
+### Test Plan
+
+In order to claim full support for this configuration, we must have
+CI coverage informing the release.
+An end-to-end job using the bootstrap-in-place installation flow,
+based on the [installer UPI CI](https://github.com/openshift/release/blob/master/ci-operator/templates/openshift/installer/cluster-launch-installer-metal-e2e.yaml#L507),
+ running an appropriate subset of the standard OpenShift tests
+will be created and configured to block accepting release images unless it passes.
+This job is a different CI from the Single node production edge CI that will run with a bootstrap vm on cloud environment.
```

Comment:
Why not single node on packet? I think it is very important that we make no assumptions regarding the instance type. Running tests on libvirt vs metal can introduce poor assumptions.

Reply:
@hexfusion it was just faster, and, of course, easier to debug. And did not require adding extra code to the installer. I guess once this work, we can add extra job using packet BM. Also, I'm not 100% sure that it's significant for u/s testing. BM tests will be done by the QE.
Hope it make sense

Reply:
But isn't this what we will ship to customers @romfreiman? If code changes are needed we need to do that regardless? Great example is with libvirt my understanding is that IP address is changing on pivot from bootstrap. This from my understanding is not how it would work in production. These nuances and corner cases are subtle but they can create a lot of work.

When a developer wants to create a SNO cluster they need to understand deployment e2e as it will be in production. If this makes debug harder then we need to consider tooling to help that as it will be the exact problem faced by support in the wild. I mean maybe this has been considered but I am not sure we should just turn a blind eye to this. I think at minimum periodic testing would need to be on metal. Do you agree? Let's not have surprises.

Reply:
@hexfusion I think that in addition to being able to run the single node CI in the cloud environment another goal here is to allow single node CI without requiring bootstrap-in-place capability from the installer.
I think we need to start with both options.
once bootstrap-in-place works (perhaps even in cloud env) and we are able to debug installation issues. the single node should use it, to test what we actually ship.

Reply:
@hexfusion I agree. Will open a ticket to add packet BM flow as well from prow regardless the packet-libvirt one.
Also, our QE will continuously run CI on bm environment which provides extra coverage. 
BTW, I assume that you're aware, all the BM ipi u/s testing is based on dev-scripts, which is packet + libvirt.

Reply:
I am fine with and understand this is where we start. I just would like to know we are exercising this on metal without assumptions.

> I assume that you're aware, all the BM ipi u/s testing is based on dev-scripts, which is packet + libvirt.

But we do a full UPI metal test on Packet. As VZN will be UPI seems reasonable to have this coverage. I think we align that a goal should be to exercise UPI metal single node directly on metal as it will be deployed in production. Appreciate the comments.

Reply:
@eranco74 or @romfreiman, do we want to add any conclusion from this thread to the enhancement?

Reply:
@hexfusion we have updated the testing section of the enhancement to justify starting with VM-based CI. Please take a look and see if that addresses your concerns.